### ISSUE TITLE
XMPP SASL implementation chokes when no realm has been received

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2098]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#2098|
|Type|defect|
|Created|2006-09-17 13:53:11Z|

As reported by Fabio Forno:

>I've noticed that same servers (e.g jabber.org) don't send the realm
>directive in the challenge and this causes the above problem in
>authenticator. At present I've solved it using the host parameter when
>no realm string is found, but I don't know if this is the correct behaviour

Exarkun then mentions:

>http://www.xmpp.org/xmppimpl.html suggests that it is.

Attachments:

* [patch.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cb2/cb23026edba0278e467358cb9801f80c00c68fc2/9c2db015e47d179fd5565f2f2144ace7e021399c.diff) (787 bytes) - added by sciasbat on 2006-09-17 21:34:34Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2098 2098
type__defect defect
reporter__ralphm ralphm
priority__highest highest
milestone__Words_0_5 Words-0.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1158501191000000 1158501191000000
changetime__1159355594000000 1159355594000000
version__None None
owner__ 
cc__ralphm cc__exarkun cc__sciasbat
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_1"></a>|@ralphm set owner to @exarkun|
|-|-|

Implemented fix in [branches/xmpp-sasl-norealm-2098.](branches/xmpp-sasl-norealm-2098.) Please review.

**Comment by twisted-trac** [comment]
|<img alt="sciasbat's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|sciasbat commented|
|-|-|

I've added a trivial patch for the issue

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm commented|
|-|-|

sciasbat: thanks, but did you miss the part where I mentioned the branch with a fix _and_ unittests?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @ralphm|
|-|-|

Changes basically look good.  I'm having some trouble doing acceptance testing.  Is there some way I can run xmpp_client.py and see it successfully authenticate against a jabber server?

I tried my lj account but only got this:

```
Connected.
SEND: <stream:stream xmlns:stream='http://etherx.jabber.org/streams' xmlns='jabber:client' to='livejournal.com' version='1.0'>
RECV: <?xml version="1.0" encoding="UTF-8"?><stream:stream from='livejournal.com' id="f71efcba2688d1f2c3a30c0da087c6e7a345a27b" version='1.0'  xmlns:stream="http://etherx.jabber.org/streams" xmlns="jabber:client"><stream:features><auth xmlns='http://jabber.org/features/iq-auth'/></stream:features>
Initialization failed.
[Failure instance: Traceback: twisted.words.protocols.jabber.xmlstream.FeatureNotAdvertized:
/home/exarkun/Projects/Twisted/trunk/twisted/internet/defer.py:191:addCallback
/home/exarkun/Projects/Twisted/trunk/twisted/internet/defer.py:182:addCallbacks
/home/exarkun/Projects/Twisted/trunk/twisted/internet/defer.py:317:_runCallbacks
/home/exarkun/Projects/Twisted/trunk/twisted/words/protocols/jabber/xmlstream.py:164:do_next
--- <exception caught here> ---
/home/exarkun/Projects/Twisted/trunk/twisted/internet/defer.py:107:maybeDeferred
/home/exarkun/Projects/Twisted/trunk/twisted/words/protocols/jabber/xmlstream.py:240:initialize
]
SEND: </stream:stream>
RECV: </stream:stream>
Disconnected.
```

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_6"></a>|@ralphm commented|
|-|-|

Ah. XMPP 1.0 compliant implementations MUST support SASL authentication, which apparently the server at lj doesn't seem to advertise. It does advertise iq-auth, which is old style authentication, per [JEP-0078](http://www.jabber.org/jeps/jep-0078.html). The lj people developed djabberd just recently and might not have gotten around implementing SASL authentication.

I made the `XMPPAuthenticator` such that it requires SASL authentication, and not do the iq-auth style authentication that is used by `BasicAuthenticator`. A hybrid would be possible, but we have recently deprecated iq-auth style authentication altogether. If this is still desirable, a separate ticket should be filed.

Can I merge this branch?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Thanks for the explanation.  Go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_8"></a>|@ralphm set status to `closed`|
|-|-|

(In [18298]) Default DIGEST-MD5 realm to host part of JID.

Author: ralphm
Reviewer: exarkun
Fixes [#2098](https://github.com/twisted/twisted/issues/2098)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
